### DISCUSSION TITLE
Apktool vs Apkzlib

### DISCUSSION THREAD

**Discussion by harshitshah4** [iBotPeaches/Apktool#2913] [Q&A]
Just wanted to understand the difference in apkzlib vs apktool.
How does it differ ?
I can see that apkzlib allows manipulating/modifying/reading contents inside zip/apk file without even need of extracting complete apk.

I can see LSPatch: https://github.com/LSPosed/LSPatch, using it for patching Android apps and seemed to be much faster too.

My use case is to just read .dex files from apk and modify it.

**Comment by iBotPeaches** [comment]
Not sure. Never heard of apkzlib and took a look. Appears to be just a good utility library for working with the `.apk` file format, which as you know is just ZIP. Has a few helper functions for dealing with the signed `.apk` and the various iterations of it, so has some handling for AndroidManifest and some of the AXML.

LSPatch uses apkzlib because it mainly runs on device is my guess so it has a different problem set than say Apktool that runs on host machine. Either way, LSPatch uses the same libraries from Android4ME that Apktool started with for resource patching.

Apktool's have changed quite heavily over the years, but LSPatch still has the originals.

Not sure what you are looking for, but hoped that helped.

**Reply by harshitshah4** [reply]
Thanks a lot for the reply, and for this amazing tool.
I have been using apktool for quite a while now and I find it pretty useful too.
But there are times where apktool fails on some apks while decompiling/compiling. That's where I was exploring apkzlib, and I figured where apktool fails, apkzlib was able to make it work. 

I am not using apkzlib for decompiling/compiling resources, but only for sources.

Just wanted to understand why apktool fails in some cases even when decompiling/compiling with sources, wanted to see If I can make some contributions to enhance it.

Also LSPatch patches the app on the host machine and then we install that app on the device.
Not sure what you meant when you mentioned that:
```
LSPatch uses apkzlib because it mainly runs on device is my guess so it has a different problem set than say Apktool that runs on host machine
```

**Reply by iBotPeaches** [reply]
> Just wanted to understand why apktool fails in some cases even when decompiling/compiling with sources, wanted to see If I can make some contributions to enhance it.

The opened 88 issues are a great place to start :)
